core/src/Cuda/Kokkos_Cuda_BlockSize_Deduction.hpp

```diff
@@ -25,17 +25,50 @@
 namespace Kokkos {
 namespace Impl {
 
+inline int cuda_warp_per_sm_allocation_granularity(cudaDeviceProp const& properties) {
+  // Allocation granularity of warps in each sm
+  switch (properties.major) {
+    case 3:
+    case 5:
+    case 7:
+    case 8:
+    case 9: return 4;
+    case 6: return (properties.minor == 0 ? 2 : 4);
+    default:
+      throw_runtime_exception("Unknown device in cuda warp per sm allocation granularity");
+      return 0;
+  }
+}
+
+inline int cuda_max_warps_per_sm_registers(cudaDeviceProp const& properties,
+                                           cudaFuncAttributes const& attributes) {
+  // Maximum number of warps per sm as a function of register counts,
+  // subject to the constraint that warps are allocated with a fixed granularity
+  int const max_regs_per_block = properties.regsPerBlock;
+  int const regs_per_warp = attributes.numRegs * properties.warpSize;
+  int const warp_granularity = cuda_warp_per_sm_allocation_granularity(properties);
+  // The granularity of register allocation is chunks of 256 registers per warp
+  int const allocated_regs_per_warp = (regs_per_warp + 256 - 1) / 256;
+
+  int const max_warps_per_sm = warp_granularity * (max_regs_per_block / (allocated_regs_per_warp * warp_granularity));
+
+  return max_warps_per_sm;
+}
+
 inline int cuda_max_active_blocks_per_sm(cudaDeviceProp const& properties,
                                          cudaFuncAttributes const& attributes,
-                                         int block_size, size_t dynamic_shmem) {
+                                         int max_warps_per_sm_registers, int block_size, size_t dynamic_shmem) {
```

Comment:
I didn't specifically benchmark it, but since it was something that was independent of `block_size` and `dynamic_shmem` it seemed like it was something good to compute once, reuse elsewhere, especially in the context of the `for` loop of the block size calculation where this function gets called multiple times.

I should have made it a `const` value, to be fair. I could also make it the last argument and give it a default value (`-1`?) such that when it isn't passed in it'll get computed inside this function. Thoughts?

Reply:
Ultimately ended up pulling the calculation in, since it is indeed cheap.